mimalloc-sys/build.rs

```diff
@@ -1,64 +1,50 @@
-//! Builds and links mimalloc
+//! Builds and links mimalloc.  Emulates behavior of CMakeLists
+// Based on flate2-rs's miniz-sys
 
-macro_rules! info {
-    ($($args:tt)*) => { println!($($args)*) }
-}
-
-use std::{env, fs, path::PathBuf};
+use std::env;
 
 fn main() {
     // Get environment variables:
     let target = env::var("TARGET").expect("TARGET was not set");
-    let host = env::var("HOST").expect("HOST was not set");
-    let profile = env::var("PROFILE").expect("PROFILE was not set");
-    let num_jobs = env::var("NUM_JOBS").expect("NUM_JOBS was not set");
-    let out_dir =
-        PathBuf::from(env::var_os("OUT_DIR").expect("OUT_DIR was not set"));
-    let src_dir = env::current_dir().expect("failed to get current directory");
-    let build_dir = out_dir.join("build");
-    let mimalloc_src_dir = src_dir.join("mimalloc");
-    let mimalloc_out_src_dir = build_dir.join("mimalloc");
 
-    info!("TARGET={}", target.clone());
-    info!("HOST={}", host.clone());
-    info!("PROFILE={}", profile.clone());
-    info!("NUM_JOBS={}", num_jobs.clone());
-    info!("OUT_DIR={:?}", out_dir);
-    info!("BUILD_DIR={:?}", build_dir);
-    info!("SRC_DIR={:?}", src_dir);
-    info!("mimalloc_src_dir={:?}", mimalloc_src_dir);
-    info!("mimalloc_out_src_dir={:?}", mimalloc_out_src_dir);
+    // Build mimalloc
+    let mut build = cc::Build::new();
+
+    build
+        .file("mimalloc/src/static.c")
+        .include("mimalloc/include")
+        .include("mimalloc/src") // For .c includes from static.c and page.c
+        .define("MI_STATIC_LIB", "1") // Tell mimalloc this is a static build
+        .warnings(false);
+
```

Comment:
The option to build a dynamic library and dynamically linked needs to be preserved somehow (ideally an environment variable should be used). 

Otherwise if you allocate memory in Rust using mimalloc, and then try to deallocate it in C, the behavior would be undefined (e.g. because two different memory allocators might be used). 

If you are going to use the static lib approach, you need to make sure that linking mimalloc statically also adds symbols for malloc, calloc, etc. (not only `mi_malloc`, `mi_calloc`, `mi_realloc`, etc.), such that if somebody attempts to link, e.g., a C library that statically or dynamically links against mimalloc they get at least a link error.

Reply:
Sorry, I'm not following why malloc must be owned by mimalloc when built statically.  So that one mimalloc datastructure would be used by both C and a rust staticlib? I can definitely see some clashes with that case.

The original build.rs had OVERRIDE forced to OFF, so I kept that behavior and didn't expose the OVERRIDE flag.  That could be an additional feature.

There's also a fun exception that just went into mimalloc from https://github.com/microsoft/mimalloc/pull/22: _malloc cannot be redefined when building static on windows_

Reply:
> There's also a fun exception that just went into mimalloc from microsoft/mimalloc#22: malloc cannot be redefined when building static on windows

Can you update the subtree to that commit ?

> The original build.rs had OVERRIDE forced to OFF, so I kept that behavior and didn't expose the OVERRIDE flag.

So I think that when linking statically we probably want to use `OVERRIDE` `ON` to force mimalloc to use the same symbols as malloc. That way, if somebody tries to link some other allocator with the same symbols they get a linker error.